README.md
# tagger-input
An HTML custom element that supports adding multiple text tags. 

[Demo](https://twitchbronbron.github.io/tagger-input/)

## Usage

```html
<tagger-input value="['Predefined', 'Values']" onchange="console.log('It changed!', event)"></tagger-input>

<tagger-input id="tagger2"></tagger-input>
<script>
    document.getElementById('tagger2').addEventListener('change', (event)=>{
        console.log('It changed', event);
    });
</script>
```

## Attributes
| Name  | Description |
|-------|-------------|
| `value`| The value of the `tagger-input` input. This can be set from html, and will also be updated as the JSON.stringify text from the `tagger-input` element |
| `onchange` | See `change` in [events](#events) table below |

## Events
| Name | Description | 
| -----|-------------|
| `change` | Emitted whenever the value has changed by the user. Will not emit when changed programmatically by setting the property or the attribute |

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2019-09-11
### Added
 - initial project release


[0.1.0]: https://github.com/TwitchBronBron/tagger-input/v0.1.0


index.js
class TaggerInputElement extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
    }

    /**
     * Add a new tag to the tag list
     */
    addTag(tagName) {
        let tagElement = document.createElement('span');
        tagElement.innerHTML = `
            <span class="tag-name">${tagName}</span>
            &nbsp;&nbsp;
            <span class="delete">
                <svg xmlns="http://www.w3.org/2000/svg" width="10" height="10" viewBox="0 0 24 24"><path d="M24 20.188l-8.315-8.209 8.2-8.282-3.697-3.697-8.212 8.318-8.31-8.203-3.666 3.666 8.321 8.24-8.206 8.313 3.666 3.666 8.237-8.318 8.285 8.203z"/></svg>
            </span>
        `;
        tagElement.getElementsByClassName('delete')[0].addEventListener('click', () => {
            this.tagContainer.removeChild(tagElement);
        });
        this.tagContainer.appendChild(tagElement);
    }

    connectedCallback() {
        this.shadowRoot.innerHTML = `
            <style>
                :host {
                    box-sizing: border-box;
                    display: block;
                    border: 1px solid rgb(238, 238, 238);
                    border-style: inset;
                    cursor: text;
                    min-height: 31px;
                }
                :host:focus{
                    outline: -webkit-focus-ring-color auto 1px;
                }
                #input {
                    border: 0;
                    outline: none;
                    display: inline-block;
                    min-width: 30px; 
                    max-width: 400px;
                    display:inline-block;
                }
                #tag-container>span {
                    display: inline-flex;
                    align-content: center;
                    justify-content: center;
                    min-width: 0;
                    padding-left: 4px;
                    padding-right: 4px;
                    border-style: solid;
                    border-width: 1px;
                    border-radius: 3px;
                    font-size: 12px;
                    line-height: 1.84615385;
                    text-decoration: none;
                    vertical-align: middle;
                    white-space: nowrap;
                    border-color: transparent;
                    background-color: #E1ECF4;
                    color: #2c5777;
                    margin: 2px;
                    font-size: 12px;
                    overflow: hidden;
                    vertical-align: middle;
                }
                .delete {
                    margin-top: 1px;
                    cursor: pointer;
                }
                .delete:hover{
                    opacity: .7;
                }
                .delete:active {
                    opacity: .9;
                }
            </style>
            <span id="tag-container">
            </span>
            <span id="input" contenteditable="true"></span>
        `;
        this.tagContainer = this.shadowRoot.getElementById('tag-container');
        this.inputElement = this.shadowRoot.getElementById('input');

        this.inputElement.addEventListener('keydown', (event) => {
            //add the new tag
            if (event.key === 'Enter') {
                this.addTag(this.inputElement.innerText.trim());
                this.inputElement.innerHTML = '';
                event.preventDefault();
                //if they click the back space at an empty text prompt, delete the last tag added
            } else if (event.key === 'Backspace' && !this.inputElement.innerText && this.tagContainer.children.length > 0) {
                this.tagContainer.removeChild(this.tagContainer.children[this.tagContainer.children.length - 1]);
            }
        });

        //focus the textbox anytime the user clicks the element
        this.addEventListener('click', (event) => {
            //only focus when clicking on the host element directly
            if (event.composedPath()[0] !== this) {
                return;
            }
            this.inputElement.focus();
            //move caret to the end
            if (window.getSelection && document.createRange) {
                var range = document.createRange();
                range.selectNodeContents(this.inputElement);
                range.collapse(false);
                var sel = window.getSelection();
                sel.removeAllRanges();
                sel.addRange(range);
            }
        });
    }
}

window.customElements.define('tagger-input', TaggerInputElement);

gulpfile.ts
import * as gulp from 'gulp';
import uglifyEs from 'gulp-uglify-es';
import * as typescript from 'gulp-typescript';
import * as replace from 'gulp-replace';
import * as rimraf from 'rimraf';
import * as fsExtra from 'fs-extra';
import * as uglifyCss from 'uglifycss';
import * as sourcemaps from 'gulp-sourcemaps';
import * as liveServer from 'live-server';


function clean(done) {
    rimraf('dist/**/*', {}, done);
}

async function getStyles() {
    let contents = await fsExtra.readFile('src/index.css');
    return uglifyCss.processString(contents.toString());
}

async function build() {
    let styles = getStyles();
    return gulp.src('src/index.ts')
        .pipe(replace('INJECT_STYLES_HERE', await styles))
        .pipe(sourcemaps.init())
        .pipe(typescript({
            target: "es2015",
            declaration: false,
            noEmitHelpers: true,
            moduleResolution: "node"
        }))
        .pipe(uglifyEs())
        .pipe(sourcemaps.write('./'))
        .pipe(gulp.dest('dist'));
}

gulp.task('watch', gulp.series(clean, build, function watch() {
    gulp.watch('src/**/*.*', build);
    liveServer.start({
        watch: [
            'dist/index.js',
            'index.html'
        ]
    });
}));
gulp.task('default', gulp.series(
    clean,
    build
));

src/index.ts
class PlumbTaggerElement extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        //create the DOM structure
        this.shadowRoot.innerHTML = `
            <span id="tags-container"></span>
            <span id="input" contenteditable="true"></span>
            <style>${PlumbTaggerElement.styles}</style>
            <template id="tag-template">
                <span class="tag">
                    <span class="tag-name"></span>
                    <span class="delete">
                        <svg xmlns="http://www.w3.org/2000/svg" width="10" height="10" viewBox="0 0 24 24"><path d="M24 20.188l-8.315-8.209 8.2-8.282-3.697-3.697-8.212 8.318-8.31-8.203-3.666 3.666 8.321 8.24-8.206 8.313 3.666 3.666 8.237-8.318 8.285 8.203z"/></svg>
                    </span>
                </span>
            </template>
        `;

        //listen for input keydown events
        this.inputElement.addEventListener('keydown', this.inputKeydown.bind(this));

        //focus the textbox anytime the user clicks the element
        this.addEventListener('click', this.hostClick.bind(this));
    }

    connectedCallback() {
        this._value = this.getValueFromAttribute() || [];

        this.renderTags();
        this.syncAttributes();
    }

    renderTags() {
        this.tagContainer.innerHTML = '';
        this._value.forEach((tagName, index) => {

            var tagElement = this.tagTemplate.content.firstElementChild.cloneNode(true) as HTMLSpanElement;
            tagElement.getElementsByClassName('tag-name')[0].innerHTML = tagName;
            tagElement.getElementsByClassName('delete')[0].addEventListener('click', () => {
                this._value.splice(index, 1);
                this.renderTags();
                this.syncAttributes();
            });
            this.tagContainer.appendChild(tagElement);
        });
    }

    /**
     * The html element where all tags get added
     */
    private get tagContainer(): HTMLSpanElement {
        return this.shadowRoot.getElementById('tags-container');
    }

    /**
     * The editable element where users will be typing
     */
    private get inputElement(): HTMLSpanElement {
        return this.shadowRoot.getElementById('input');
    }

    private get tagTemplate(): HTMLTemplateElement {
        return this.shadowRoot.getElementById('tag-template') as any;
    }

    private static styles = `INJECT_STYLES_HERE`;

    get value() {
        return this._value;
    }
    set value(value) {
        this._value = Array.isArray(value) ? value : [];
        //draw all the tags
        this.renderTags();
        //update the attribute with this value
        this.syncAttributes();
    }
    private _value: string[];

    public static get observedAttributes() {
        return ['value'];
    }

    attributeChangedCallback(name, oldValue, newValue) {
        if (this.tagContainer && oldValue !== newValue) {
            if (name === 'value') {
                let value = this.getValueFromAttribute();
                if (Array.isArray(value)) {
                    this._value = value;
                    this.renderTags();
                    this.syncAttributes();
                }
            }
        }
    }

    /**
     * Synchronizes the attributes with the state
     */
    private syncAttributes() {
        //value attribute
        let currentValue = this.getAttribute('value');
        let newValue = JSON.stringify(this._value)
        if (currentValue !== newValue) {
            this.setAttribute('value', newValue);
        }

        //empty attribute
        if (this._value.length === 0) {
            this.setAttribute('empty', '');
        } else {
            this.removeAttribute('empty');
        }
    }

    private getValueFromAttribute() {
        const attr = this.getAttribute('value');
        const attrValue = eval(`(function(){return ${attr};})()`);
        if (Array.isArray(attrValue)) {
            return attrValue;
        } else {
            return undefined;
        }
    }

    /**
     * Fires any time the input element has a keydown press
     * @param event 
     */
    private inputKeydown(event) {
        //add the new tag
        if (event.key === 'Enter') {
            //get the tag from the input element
            let tag = this.inputElement.innerText.trim();
            if (tag.length > 0) {
                //clear the input element
                this.inputElement.innerText = '';

                //save the tag
                this._value.push(tag);

                this.renderTags();
                this.syncAttributes();

                this.emitEvent(new CustomEvent('change', {
                    bubbles: false,
                    detail: this._value
                }));
            }
            event.preventDefault();
            //if they click the back space at an empty text prompt, delete the last tag added
        } else if (event.key === 'Backspace' && !this.inputElement.innerText && this.tagContainer.children.length > 0) {
            this._value.pop();
            this.renderTags();
            this.syncAttributes();

            this.emitEvent(new CustomEvent('change', {
                bubbles: false,
                detail: this._value
            }));
        }
    }

    private emitEvent(event: CustomEvent) {
        //dispatch the event the normal way
        this.dispatchEvent(event);

        //try to call an attribute event handler if one exists
        let attr = this.getAttribute(`on${event.type}`);
        if (attr) {

            //run the code in the attribute. If it's NOT a function instance, this is the actual action
            let result = eval(`(function(event){return ${attr};})(event)`);

            //if we received a function back, assume we are supposed to call that function, so pass the event to it as its first argument
            if (typeof result === 'function') {
                result(event);
            }
        }
    }

    /**
     * Called any time the host element is clicked
     * @param event
     */
    private hostClick(event) {
        //only focus when clicking on the host element directly
        if (event.composedPath()[0] !== this) {
            return;
        }
        this.inputElement.focus();
        //move caret to the end
        if (window.getSelection && document.createRange) {
            var range = document.createRange();
            range.selectNodeContents(this.inputElement);

            range.collapse(false);
            var sel = window.getSelection();
            sel.removeAllRanges();
            sel.addRange(range);
        }
    }
}

window.customElements.define('tagger-input', PlumbTaggerElement);